just_audio/ios/Classes/FFTHelper.m

```diff
@@ -0,0 +1,67 @@
+//
+//  FFTHelper.c
+//  just_audio
+//
+//  Created by Eittipat on 10/10/2564 BE.
+//
+//  Original source codes
+//  - https://developer.android.com/reference/android/media/audiofx/Visualizer#getFft(byte[])
+//  - http://www.myuiviews.com/2016/03/04/visualizing-audio-frequency-spectrum-on-ios-via-accelerate-vdsp-fast-fourier-transform.html
```

Comment:
I've been stuck on just the the legal side of things with copying this code, and the code you have linked to is not actually licensed (i.e. no permission is granted to copy it.)

I will try to contact the original author and ask him for permission, otherwise it would be necessary to write this clean.

Reply:
I know how you feel, but copyright is a little tricky. It is better to get permission (which I'm in the process of getting), unless you were to write it clean without looking at the other code. How it gets complicated is that copyright protects not the idea but the expression of the idea. So for example, it could be argued that it would be very unlikely for you to have chosen the variable names that you did were not not for following the original code. And as for whether you could simply rename all the variables, the issue here is that there is now a historical record of how you arrived at the end result, which started with a copy and then making superficial changes (in the court of law they would also look at whether it could be proved that you had access to the original work - although let me just say as a disclaimer that IANAL). So anyway, yes to ensure that everything is OK it is always best to cross the t's and dot the i's in legal matters, and companies who link just_audio into their apps would also want be able to trust that all code is legitimately licensed.

Personally when I read a tutorial, I like to first check if the code has a license, and if not, I will read the tutorial only to learn, but then try to write my own code based only on the understanding I gained from the article rather than by constantly looking to the next line of code in their example before writing my own. In any case, it should be hopefully resolved by contacting the original author and just getting permission.